snuba_sdk/query_visitors.py

```diff
@@ -0,0 +1,321 @@
+        if select is not None:
+            return f"SELECT {', '.join(self.translator.visit(s) for s in select)}"
+        return ""
+
+    def _visit_groupby(self, groupby: Optional[List[Union[Column, Function]]]) -> str:
+        if groupby is not None:
+            return f"BY {', '.join(self.translator.visit(g) for g in groupby)}"
+        return ""
+
+    def _visit_where(self, where: Optional[List[Condition]]) -> str:
+        if where is not None:
+            return f"WHERE {' AND '.join(self.translator.visit(w) for w in where)}"
+        return ""
+
+    def _visit_having(self, having: Optional[List[Condition]]) -> str:
+        if having is not None:
+            return f"HAVING {', '.join(self.translator.visit(h) for h in having)}"
+        return ""
+
+    def _visit_orderby(self, orderby: Optional[List[OrderBy]]) -> str:
+        if orderby is not None:
+            return f"ORDER BY {', '.join(self.translator.visit(o) for o in orderby)}"
+        return ""
+
+    def _visit_limitby(self, limitby: Optional[LimitBy]) -> str:
+        if limitby is not None:
+            return f"LIMIT {self.translator.visit(limitby)}"
+        return ""
+
+    def _visit_limit(self, limit: Optional[Limit]) -> str:
+        if limit is not None:
+            return f"LIMIT {self.translator.visit(limit)}"
+        return ""
+
+    def _visit_offset(self, offset: Optional[Offset]) -> str:
+        if offset is not None:
+            return f"OFFSET {self.translator.visit(offset)}"
+        return ""
+
+    def _visit_granularity(self, granularity: Optional[Granularity]) -> str:
+        if granularity is not None:
+            return f"GRANULARITY {self.translator.visit(granularity)}"
+        return ""
+
+    def _visit_totals(self, totals: Optional[Totals]) -> str:
+        if totals is not None:
+            return f"TOTALS {self.translator.visit(totals)}"
+        return ""
+
+
+class Translator(Printer):
+    def __init__(self) -> None:
+        super().__init__(False)
+
+    def _combine(self, query: Query, returns: MutableMapping[str, str]) -> str:
+        formatted_query = super()._combine(query, returns)
+        body = {"dataset": query.dataset, "query": formatted_query}
+
+        return json.dumps(body)
+
+
+class Validator(QueryVisitor[None]):
+    def _combine(self, query: Query, returns: MutableMapping[str, None]) -> None:
+        # TODO: Contextual validations:
+        # - Must have certain conditions (project, timestamp, organization etc.)
+
+        if query.select is None:
+            raise InvalidQuery("query must have at least one column in select")
+
+        # - Orderby must be a field in select
+        if query.orderby is not None:
+            for o in query.orderby:
+                found = False
+                for s in query.select:
+                    if s == o.exp:
+                        found = True
+                        break
```

Comment:
Why do we enforce that the order by is in the select ? A query that sorts according to a key and then return other fields does not seem necessarily invalid to me.

Also I don't think this works. 
If I have `SELECT a, b ,c`
and `ORDER BY a, x, c`

If you want to ensure all columns in the order by are also in the select, this check will pass while it should not.

Reply:
That's my mistake I thought that would be invalid but you're right clickhouse allows ordering by non-selected fields. I'll remove this check.